Add I18n and L10n tips and tricks (#129)
infinum/eightshift-docs: website/docs/basics/tips-tricks.md

	}
}
```

### Internationalization (I18n) and localization (L10n)

> To make sure your project is translatable into other languages, you should [WordPress's i18n functions](https://codex.wordpress.org/I18n_for_WordPress_Developers) every time you output text. This process, called internationalization, is quite simple to implement by using the `__` WordPress function, which is available in both PHP and Gutenberg (in the `@wordpress/i18n` package, and is standard practice in Eightshift Frontend Libs for all text output. The `__` function accepts a string to translate and a text domain, which is used to tell WordPress which translation set to use, and returns the translated string (or original string, in case a translation doesn't exist). Note that you can provide a `sprintf`-parametrized function to `__` and use `sprintf` on its return value to populate the string with certain values. You should also load the text domain using `load_theme_textdomain`. Using the Eightshift Libs `I18n` class will do this for you by instructing WordPress to look for translations in `src/I18n/languages` with the text domain defined as your project name.

Reviewer: No idea why it's not highlighted, but anyway, I added "...you should `use` WordPress's i18n functions..." in the first sentence

> To make sure your project is translatable into other languages, you should [WordPress's i18n functions](https://codex.wordpress.org/I18n_for_WordPress_Developers) every time you output text. This process, called internationalization, is quite simple to implement by using the `__` WordPress function, which is available in both PHP and Gutenberg (in the `@wordpress/i18n` package, and is standard practice in Eightshift Frontend Libs for all text output. The `__` function accepts a string to translate and a text domain, which is used to tell WordPress which translation set to use, and returns the translated string (or original string, in case a translation doesn't exist). Note that you can provide a `sprintf`-parametrized function to `__` and use `sprintf` on its return value to populate the string with certain values. You should also load the text domain using `load_theme_textdomain`. Using the Eightshift Libs `I18n` class will do this for you by instructing WordPress to look for translations in `src/I18n/languages` with the text domain defined as your project name.

Reviewer 2: There is also unclosed `)` in the `(in the `@wordpress/i18n` package,` -> `(in the `@wordpress/i18n` package),`

> To make sure your project is translatable into other languages, you should [WordPress's i18n functions](https://codex.wordpress.org/I18n_for_WordPress_Developers) every time you output text. This process, called internationalization, is quite simple to implement by using the `__` WordPress function, which is available in both PHP and Gutenberg (in the `@wordpress/i18n` package, and is standard practice in Eightshift Frontend Libs for all text output. The `__` function accepts a string to translate and a text domain, which is used to tell WordPress which translation set to use, and returns the translated string (or original string, in case a translation doesn't exist). Note that you can provide a `sprintf`-parametrized function to `__` and use `sprintf` on its return value to populate the string with certain values. You should also load the text domain using `load_theme_textdomain`. Using the Eightshift Libs `I18n` class will do this for you by instructing WordPress to look for translations in `src/I18n/languages` with the text domain defined as your project name.

Reviewer 2: I'd also put a break line in 

```
 The `__` function
```

part, for easier readability.

> To make sure your project is translatable into other languages, you should [WordPress's i18n functions](https://codex.wordpress.org/I18n_for_WordPress_Developers) every time you output text. This process, called internationalization, is quite simple to implement by using the `__` WordPress function, which is available in both PHP and Gutenberg (in the `@wordpress/i18n` package, and is standard practice in Eightshift Frontend Libs for all text output. The `__` function accepts a string to translate and a text domain, which is used to tell WordPress which translation set to use, and returns the translated string (or original string, in case a translation doesn't exist). Note that you can provide a `sprintf`-parametrized function to `__` and use `sprintf` on its return value to populate the string with certain values. You should also load the text domain using `load_theme_textdomain`. Using the Eightshift Libs `I18n` class will do this for you by instructing WordPress to look for translations in `src/I18n/languages` with the text domain defined as your project name.

Reviewer 2: ```suggestion
To make sure your project is translatable into other languages, you should use the [WordPress's i18n functions](https://codex.wordpress.org/I18n_for_WordPress_Developers) every time you output text. This process, called internationalization, is quite simple to implement by using the `__` WordPress function, which is available in both PHP and Gutenberg (in the `@wordpress/i18n` package), and is standard practice in Eightshift Frontend Libs for all text output. 

The `__` function accepts a string to translate and a text-domain, which is used to tell WordPress which translation set to use and returns the translated string (or original string, in case the translation doesn't exist). Note that you can provide a `sprintf`-parametrized function to `__` and use `sprintf` on its return value to populate the string with certain values. You should also load the text domain using the `load_theme_textdomain` function. Using the Eightshift Libs `I18n` class will do this for you by instructing WordPress to look for translations in the `src/I18n/languages` folder with the text domain defined as your project name.
```

> However, there will probably come a time when _being able_ to translate won't be enough. You'll actually need to _do_ that. This happens in a process called localization, and it's a bit less trivial to do.

Reviewer 2: ```suggestion
However, there will come a time when _being able_ to translate won't be enough. You'll actually need to _do_ that. This happens in a process called localization and is a bit less trivial to do.
```

> First you'll need to prepare the strings for translators. As WordPress i18n is based upon `gettext`, you should generate a POT ("portable object template") file which you can do by running WP-CLI command `wp i18n make-pot` (https://developer.wordpress.org/cli/commands/i18n/make-pot/) in your project root. Note that you may need to exclude vendor folders.

Reviewer 2: ```suggestion
First, you'll need to prepare the strings for translators. As WordPress i18n is based upon the `gettext` system, you should generate a POT ("portable object template") file which you can do by running WP-CLI command `wp i18n make-pot` (https://developer.wordpress.org/cli/commands/i18n/make-pot/) in your project root. Note that you may need to exclude vendor folders.
```

> You can provide this POT file to translators directly, and they can use tools for PO editing such as Poedit to translate the strings. After that, they can export an MO file which you can import into your project. Alternatively, you can set up a translation environment like Glotpress for translations, import your POT file there and export translations as an PO and MO file. You should locate these files into the path defined in your `load_theme_textdomain` call, and they should have the name `{textdomain}-{locale}.mo` (e.g. `eightshift-boilerplate-hr_HR.mo`), that is, `{textdomain}-{locale}.po`.

Reviewer 2: ```suggestion
You can provide this POT file to translators directly, and they can use tools for PO editing such as Poedit to translate the strings. After that, they can export an MO file which you can import into your project. Alternatively, you can set up a translation environment like Glotpress for translations, import your POT file there, and export translations as PO and MO files. You should locate these files into the path defined in your `load_theme_textdomain` call, and they should have the name `{textdomain}-{locale}.mo` (e.g. `eightshift-boilerplate-hr_HR.mo`), that is, `{textdomain}-{locale}.po`.
```

> However, MO translations are only loaded by PHP code, and aren't understood by Gutenberg. To do so, you'll need to generate a JED file from the PO file. You can probably use the WP CLI `wp i18n make-json` file for that (https://developer.wordpress.org/cli/commands/i18n/make-json/). If you're using Glotpress, you can simply export the locale as a JED file.

Reviewer 2: ```suggestion
MO translations are only loaded by PHP code and aren't understood by Gutenberg. To do so, you'll need to generate a JED file from the PO file. You can probably use the WP CLI `wp i18n make-json` file for that (https://developer.wordpress.org/cli/commands/i18n/make-json/). If you're using Glotpress, you can simply export the locale as a JED file.
```

> To instruct Gutenberg on how to load your JED translations, you'll need to call `wp_set_script_translations` providing a text domain, handle (script identifier), and locale path. If you're using the Eightshift Libs `I18n` class, this is done for you with the same assumptions as for PHP translations. Furthermore, as we're assuming you're setting translations for block editor scripts, we set the handle to `{$assetsPrefix}-block-editor-scripts`. If you haven't changed how scripts are enqueued, this should be the correct handle and your assets prefix should be your project name.

Reviewer 2: ```suggestion
To instruct Gutenberg on how to load your JED translations, you'll need to call `wp_set_script_translations` providing a text-domain, handle (script identifier), and locale path. If you're using the Eightshift Libs `I18n` class, this is done for you with the same assumptions as for PHP translations. Furthermore, as we're assuming you're setting translations for block editor scripts, we set the handle to `{$assetsPrefix}-block-editor-scripts`. If you haven't changed how scripts are enqueued, this should be the correct handle and your assets prefix should be your project name.
```
